vortexor/src/vortexor.rs

```diff
@@ -0,0 +1,216 @@
+
+pub struct TpuSockets {
+    pub tpu_quic: Vec<UdpSocket>,
+    pub tpu_quic_fwd: Vec<UdpSocket>,
+}
+
+pub struct TpuStreamerConfig {
+    pub tpu_thread_name: &'static str,
+    pub tpu_metrics_name: &'static str,
+    pub tpu_fwd_thread_name: &'static str,
+    pub tpu_fwd_metrics_name: &'static str,
+    pub quic_server_params: QuicServerParams,
+}
+
+pub struct Vortexor {
+    thread_handles: Vec<JoinHandle<()>>,
+    key_update_notifier: Arc<KeyUpdateNotifier>,
+}
+
+struct KeyUpdateNotifier {
+    key_updaters: Mutex<Vec<Arc<EndpointKeyUpdater>>>,
+}
+
+impl KeyUpdateNotifier {
+    fn new(key_updaters: Vec<Arc<EndpointKeyUpdater>>) -> Self {
+        Self {
+            key_updaters: Mutex::new(key_updaters),
+        }
+    }
+}
+
+impl NotifyKeyUpdate for KeyUpdateNotifier {
+    fn update_key(&self, key: &Keypair) -> Result<(), Box<dyn std::error::Error>> {
+        let updaters = self.key_updaters.lock().unwrap();
+        for updater in updaters.iter() {
+            updater.update_key(key)?
+        }
+        Ok(())
+    }
+}
+
+impl Vortexor {
+    pub fn create_tpu_sockets(
+        bind_address: std::net::IpAddr,
+        dynamic_port_range: (u16, u16),
+        num_quic_endpoints: u64,
+    ) -> TpuSockets {
+        let quic_config = SocketConfig { reuseport: true };
+
+        let (_, tpu_quic) =
+            bind_in_range_with_config(bind_address, dynamic_port_range, quic_config.clone())
+                .expect("expected bind to succeed");
+
+        let tpu_quic_port = tpu_quic.local_addr().unwrap().port();
+        let tpu_quic = bind_more_with_config(
+            tpu_quic,
+            num_quic_endpoints.try_into().unwrap(),
+            quic_config.clone(),
+        )
+        .unwrap();
+
+        let (_, tpu_quic_fwd) = bind_in_range_with_config(
+            bind_address,
+            (tpu_quic_port.saturating_add(1), dynamic_port_range.1),
+            quic_config.clone(),
+        )
+        .expect("expected bind to succeed");
+
+        let tpu_quic_fwd = bind_more_with_config(
+            tpu_quic_fwd,
+            num_quic_endpoints.try_into().unwrap(),
+            quic_config,
+        )
+        .unwrap();
+
+        TpuSockets {
+            tpu_quic,
+            tpu_quic_fwd,
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn create_vortexor(
+        tpu_sockets: TpuSockets,
+        staked_nodes: Arc<RwLock<StakedNodes>>,
+        tpu_sender: Sender<PacketBatch>,
+        tpu_fwd_sender: Sender<PacketBatch>,
+        max_connections_per_peer: u64,
+        max_tpu_staked_connections: u64,
+        max_tpu_unstaked_connections: u64,
+        max_streams_per_ms: u64,
+        max_connections_per_ipaddr_per_min: u64,
+        tpu_coalesce: Duration,
+        identity_keypair: &Keypair,
+        exit: Arc<AtomicBool>,
+    ) -> Self {
+        let config = TpuStreamerConfig {
+            tpu_thread_name: "solQuicTpu",
+            tpu_metrics_name: "quic_streamer_tpu",
```

Comment:
This will collide with the metrics name for the TPU streamer in main validator. Can we rename it?

Reply:
We can -- I was torn between using the quic_streamer_tpu and having a new metric name. There is pro and cons. The pro sharing the name is to it is easier to look at the network wide metrics -- after all they are ingesting transactions the same as regular validator and metrics are relevant. The con is it is harder to look at vortexor specific metrics.

Reply:
The bigger con in when there's a hybrid network, where only some nodes are using vortexor. It'll be hard to differentiate between the metrics.

Reply:
agreed